Hw02 express

Alex-T-T opened the pull request:
done Hw02 express

Alex-T-T committed:
first steps

Alex-T-T committed:
some changes

Alex-T-T committed:
without update

Alex-T-T committed:
add updateContact

Alex-T-T committed:
add validation

1 add folders "src", "middleware"
2 change position of folders "models", "routes" to "src"
3 add validation for POST & PUT routes

Alex-T-T committed:
add controllers

1 add folder "controllers"
2 create file contactControllers
3 replace logic of routes to contactControllers
4 change imports

Alex-T-T committed:
eslint fix

OnyshchenkoO commented on .eslintrc.js:
```diff
@@ -1,12 +1,15 @@
+        "node": true
+    },
+    "extends": [
+        "google"
```
в README вказано, що має використовуватись Prettier

Alex-T-T replied:
+ змінив

OnyshchenkoO commented on app.js:
```diff
@@ -1,25 +1,25 @@
-app.use(cors())
-app.use(express.json())
+app.use(logger(formatsLogger));
+// app.use(cors())
```
Чого ти це закоментував?

Alex-T-T replied:
бо як я зрозумів, то в цій ДЗ ми не використовуємо дану технологію та не торкаємося headers в запитах.
вже розкоментував

OnyshchenkoO commented on src/controllers/contactsController.js:
```diff
@@ -0,0 +1,62 @@
+const putContact = async (req, res, next) => {
+  const {name, email, phone} = req.body;
+
+  if (!name && !email && !phone) {
```
Ця перевірка теж відноситься до валідації.  Я б переніс її в файл src/middlewares/validationMiddlware.js одразу після 33 рядка
І б радив робити перевірку Object.keys(req.body).length === 0 , бо якщо в тебе зміниться модель, ти можеш забути , що ти деструктуризуєш тут властивності.

Alex-T-T replied:
готово

OnyshchenkoO commented on src/models/contacts.js:
```diff
@@ -0,0 +1,74 @@
+  return newContact;
+};
+
+const updateContact = async (contactId, body) => {
+  const {name, email, phone} = body;
+
+  const contacts = await listContacts();
+  const contactIndex = contacts.findIndex((contact) =>
+    contact.id === contactId.toString());
+
+  if (contactIndex === -1) {
+    return null;
+  };
+  const updatedContact = {id: contactId, name, email, phone};
+
+  const newContactList = contacts.filter((contact) =>
+    contact.id !== contactId.toString());
+  newContactList.push(updatedContact);
+  await fs.writeFile(contactsPath, JSON.stringify(newContactList));
+  return updatedContact;
+};
```
У тебе неправильно реалізована ця функція.

В тебе body може мати лише name, того коли ти робиш updatedContact, в тебе буде email i phone будуть undefined

Alex-T-T replied:
Функцію змінив, сподіваюсь, що правильно.
Але є запитання: 
ʼʼВ тебе body може мати лише name, того коли ти робиш updatedContact, в тебе буде email i phone будуть undefinedʼʼ - при такому підході - в попередній версії функції все рівно відбувався update  контакту і ніяких помилок не було. Чому так?

OnyshchenkoO replied:
Яке у тебе було тіло запиту в постмені коли ти робив апдейт?

Alex-T-T replied:
робив кілька запитів, в тому числі і з одним полем (імені чи номеру )

OnyshchenkoO replied:
![image](https://user-images.githubusercontent.com/48943732/205093829-10e82869-8dc8-4558-adfe-38b113591769.png)

Відкодився до коміту, який тоді був і не схоже, що воно працювало)

Alex-T-T replied:
https://drive.google.com/file/d/1KdEdAVSOJATPAz2in3xxpcaR_Ifc00iT/view?usp=share_link

ось код тієї функції і він працює - і записує апдейт контакту

Alex-T-T replied:
я так зрозумів, що треба підшаманити валідацію, щоб імʼя могло бути лише у форматі a-z, A-Z, без цифр!?

OnyshchenkoO replied:
По валідації - не так критично, як душа бажає

OnyshchenkoO commented on src/controllers/contactsController.js:
```diff
@@ -0,0 +1,62 @@
+  res.status(200).json({contacts, status: 200});
+};
+
+const getOneContactById = async (req, res, next) => {
```
Краще назвати getContactById - бо в тебе Contact, того й так зрозуміло, шо повернеться 1 контакт

Alex-T-T replied:
назву функції змінив, але в даному випадку напевно компромісним рішенням буде getContact, бо назва функція   getContactById вже існує у файлі src/models/contacts.js і імпортується у вказаний файл

Alex-T-T committed:
lit fix

Alex-T-T committed:
uncomment cors

uncomment cors at app.js

Alex-T-T committed:
update validation

1. replace checking req.body from contactsControllers.js to validationMiddleware.js
2. update checking req.body

Alex-T-T committed:
rename function for GET

getOneContactById renamed

Alex-T-T committed:
fix updateContact

Alex-T-T committed:
contacts.json

change update when fs.writeFile is done

OnyshchenkoO commented on src/models/contacts.js:
```diff
@@ -0,0 +1,70 @@
+        if (contactIndex === -1) {
+        return null;
+        };
+        contacts[contactIndex] = { id: contactId, ...body };
```
Ти робиш те саме, що робив в минулій версії. Тобі треба взяти існуючий контакт і оновити його відповідно до новий властивостей, що прийдуть в body

Alex-T-T replied:
в лекціях повсюди кажуть, що PUT використовується для повної заміни елемента, а PATCH - для часткового оновлення, та й на прикладах так у наших лекціях показують (тому я собі й думав, що треба повністю перезаписувати контакт, не зважаючи на те, що в наявному контакті та скільки полів приходить у body, і відповідно передавав лише новий контакт зі старим ID). -\_(!_!)_/-
 
Я правильно зрозумів: 
1. отримую список  контактів 
2. знаходжу індекс потрібного елемента (за ID)
3. створюю новий обʼєкт, в який розпилюю наявний контакт та наявні дані, що прийшли
4. видаляю з масиву наявний контакт (повертаю новий масив без контакту  (за ID)
5. додаю до масиву новий контакт (створений в п.3)
6. записую новий масив до contacts.json

OnyshchenkoO replied:
Так, і є. PUT використовується для заміни, якщо існує елемент, або створення - якщо ні. Це завдання неправильно написано, я б там краще PATCH ендпоінт вимагав написати. Бо 2 пунки вимагає будь-які поля моделі, що більше підходить під PATCH, бо для PUT треба кидати всі поля моделі
Так й для навчання PATCH кращий, бо треба подумати, а в PUT - те саме, що ви для POST робите

Я так реалізував цей сервіс

```
const updateContact = async (contactId, body) => {
  try {
    const existingContacts = await listContacts()
    const contactIndex = existingContacts.findIndex((row => +row.id === +contactId))

    const updatedContact = {...existingContacts[contactIndex], ...body}

    existingContacts[contactIndex] = updatedContact

    await fs.writeFile(contactsPath, JSON.stringify(existingContacts))
    return updatedContact
  } catch (err) {
    console.log(err)
  }
}
```

Alex-T-T replied:
`        const updatedContact = { ...contacts[contactIndex], ...body }
        const newContactList = contacts.filter((contact) =>
        contact.id !== contactId.toString());
        newContactList.push(updatedContact)
        await fs.writeFile(contactsPath, JSON.stringify(newContactList, null, 4)); 
        return updatedContact;`

ну я майже те ж саме зробив, але прямо написав, щоб видалити старий контакт та додати контакт, що проапдейтили в масив.

якщо це критично, то зміню як у тебе

Alex-T-T replied:
хоча виходить, що критично, бо по моєму варіанту завжди елемент буде додаватися в кінець масиву 
і в нього буде новий індекс

Alex-T-T committed:
fix validation

fix validation for name & phone number

Alex-T-T committed:
fix updateContact

Alex-T-T committed:
once more fix updateContact
